File: apps/dzik-os/frontend/src/dates.ts
// Wspólny moduł dat (jedno źródło prawdy — patrz backend/dzik_os/dates.py).
//
// Model dat w aplikacji:
// 1. Data kalendarzowa użytkownika (performed_on, logged_on, occurred_on,
//    completed_on, week_start, measured_at, due_date, target_date):
//    "YYYY-MM-DD" wyliczana w LOKALNEJ strefie przeglądarki. NIGDY przez
//    `new Date().toISOString().slice(0, 10)` — to data UTC, która o 00:30
//    czy 01:00 czasu polskiego wskazuje jeszcze wczorajszy dzień.
// 2. Dokładny moment zdarzenia (created_at, paid_at, read_at...): pełny
//    timestamp UTC z backendu; do strefy lokalnej przeliczany dopiero przy
//    prezentacji (plDateTime).
// 3. Termin lokalny (consult_slots.starts_at): naiwny "YYYY-MM-DDTHH:MM"
//    w strefie lokalnej; porównywany wyłącznie z localNowMinute().
//
// Parametr `now` służy testom — funkcje są czyste względem niego.

const pad2 = (n: number) => String(n).padStart(2, "0");

/** Data kalendarzowa Date → "YYYY-MM-DD" w lokalnej strefie. */
export function toIsoDate(d: Date): string {
  return `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`;
}

/** Dzisiejsza data kalendarzowa w lokalnej strefie ("YYYY-MM-DD"). */
export const localToday = (now: Date = new Date()): string => toIsoDate(now);

/** Poniedziałek tygodnia zawierającego `now` ("YYYY-MM-DD", lokalnie). */
export function mondayOfWeek(now: Date = new Date()): string {
  const d = new Date(now.getFullYear(), now.getMonth(), now.getDate());
  const day = d.getDay() || 7; // niedziela=0 → 7
  d.setDate(d.getDate() - day + 1);
  return toIsoDate(d);
}

/** Lokalny czas z dokładnością do minuty ("YYYY-MM-DDTHH:MM") — jedyny
 * poprawny komparand dla terminów konsultacji (starts_at). */
export const localNowMinute = (now: Date = new Date()): string =>
  `${toIsoDate(now)}T${pad2(now.getHours())}:${pad2(now.getMinutes())}`;

/** Parsuje wartość z API: "YYYY-MM-DD" jako LOKALNĄ północ (data
 * kalendarzowa nie jest momentem — parsowanie przez `new Date(iso)`
 * dałoby północ UTC i ryzyko przesunięcia dnia), pozostałe formaty
 * (pełne timestampy UTC, naiwne "YYYY-MM-DDTHH:MM") natywnie. */
export function parseApiDate(iso: string): Date {
  const m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(iso);
  if (m) return new Date(Number(m[1]), Number(m[2]) - 1, Number(m[3]));
  return new Date(iso);
}

/** Prezentacja daty kalendarzowej po polsku. */
export const plDate = (iso: string | null | undefined) =>
  iso
    ? parseApiDate(iso).toLocaleDateString("pl-PL", { day: "numeric", month: "long", year: "numeric" })
    : "—";

/** Prezentacja momentu zdarzenia po polsku (przeliczenie do strefy
 * lokalnej następuje dopiero tutaj). */
export const plDateTime = (iso: string | null | undefined) =>
  iso
    ? parseApiDate(iso).toLocaleString("pl-PL", { day: "numeric", month: "short", hour: "2-digit", minute: "2-digit" })
    : "—";

export const WEEKDAYS = ["pon", "wt", "śr", "czw", "pt", "sob", "niedz"];

File: apps/dzik-os/backend/dzik_os/dates.py
"""Wspólny moduł dat i stref czasowych (jedno źródło prawdy).

Model dat w aplikacji (patrz też frontend/src/dates.ts):

1. **Data kalendarzowa użytkownika** — `performed_on`, `logged_on`,
   `occurred_on`, `completed_on`, `week_start`, `measured_at` (pomiary),
   `due_date`, `start_date`/`end_date`, `target_date`, `taken_at` (zdjęcia):
   format ``YYYY-MM-DD`` wyliczony w LOKALNEJ strefie użytkownika
   (frontend: strefa przeglądarki; backend: `tz_for_user()`). Nigdy nie
   wyliczać jej przez ``datetime.now(UTC).date()`` — o 00:30 czasu
   polskiego dałoby to wczorajszą datę.

2. **Dokładny moment zdarzenia** — `created_at`, `updated_at`, `paid_at`,
   `read_at`, `booked_at`, `granted_at`, łańcuch audytu: pełny timestamp
   UTC (`models.now_iso()`); do strefy lokalnej przeliczany dopiero przy
   prezentacji (frontend `plDateTime`).

3. **Termin lokalny** — `consult_slots.starts_at`: naiwny
   ``YYYY-MM-DDTHH:MM`` w strefie `tz_for_user()`; porównywany wyłącznie
   z `local_now_minute()` (nigdy z czasem UTC).

4. **Data rozliczeniowa** — `payment_records.due_date`: jak (1); zaległość
   ("overdue") liczona względem `local_today()`.

Parametr ``user`` to punkt rozszerzenia pod strefę per użytkownik —
przekazuj obiekt `User` tam, gdzie jest pod ręką, a gdy w przyszłości
zyska pole `timezone`, wystarczy zmiana w `tz_for_user()`.
Parametr ``now`` (świadomy datetime, zwykle UTC) służy testom.
"""

from __future__ import annotations

from datetime import UTC, date, datetime
from zoneinfo import ZoneInfo

from .config import settings


def tz_for_user(user: object | None = None) -> ZoneInfo:
    """Strefa czasowa użytkownika.

    Od migracji nr 14 model `User` ma pole `timezone` (IANA, ustawiane w
    ustawieniach powiadomień); NULL lub brak obiektu = strefa aplikacji
    (DZIK_TZ, domyślnie Europe/Warsaw). Nieznana nazwa strefy w bazie nie
    może wywrócić żądania — fallback do strefy aplikacji.
    """
    tz_name = getattr(user, "timezone", None) or settings.timezone
    try:
        return ZoneInfo(tz_name)
    except Exception:  # noqa: BLE001 - uszkodzona wartość nie psuje aplikacji
        return ZoneInfo(settings.timezone)


def local_now(user: object | None = None, *, now: datetime | None = None) -> datetime:
    """Bieżący czas w lokalnej strefie użytkownika (świadomy datetime)."""
    instant = now if now is not None else datetime.now(UTC)
    return instant.astimezone(tz_for_user(user))


def local_today(user: object | None = None, *, now: datetime | None = None) -> date:
    """Dzisiejsza data kalendarzowa w lokalnej strefie użytkownika."""
    return local_now(user, now=now).date()


def local_today_iso(user: object | None = None, *, now: datetime | None = None) -> str:
    """`local_today()` jako ``YYYY-MM-DD`` (format API/bazy)."""
    return local_today(user, now=now).isoformat()


def local_now_minute(user: object | None = None, *, now: datetime | None = None) -> str:
    """Lokalny czas z dokładnością do minuty (``YYYY-MM-DDTHH:MM``) —
    jedyny poprawny komparand dla `consult_slots.starts_at`."""
    return local_now(user, now=now).strftime("%Y-%m-%dT%H:%M")


def parse_iso_date(value: str) -> date:
    """Parsuje datę kalendarzową ``YYYY-MM-DD`` (bez strefy — data nie jest
    momentem). Podnosi ValueError dla nieprawidłowego wejścia."""
    return date.fromisoformat(value[:10])
